File: python/pyautd3/link/twincat.py
"""
File: twincat.py
Project: link
Created Date: 21/10/2022
Author: Shun Suzuki
-----
Last Modified: 27/10/2023
Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
-----
Copyright (c) 2022-2023 Shun Suzuki. All rights reserved.

"""

from datetime import timedelta

from pyautd3.internal.link import Link, LinkBuilder
from pyautd3.internal.utils import _validate_ptr
from pyautd3.native_methods.autd3capi import (
    NativeMethods as Base,
)
from pyautd3.native_methods.autd3capi_def import ControllerPtr, LinkBuilderPtr, LinkPtr
from pyautd3.native_methods.autd3capi_link_twincat import LinkRemoteTwinCATBuilderPtr, LinkTwinCATBuilderPtr
from pyautd3.native_methods.autd3capi_link_twincat import NativeMethods as LinkTwinCAT


class TwinCAT(Link):
    """Link using TwinCAT3."""

    class _Builder(LinkBuilder):
        _builder: LinkTwinCATBuilderPtr

        def __init__(self: "TwinCAT._Builder") -> None:
            self._builder = LinkTwinCAT().link_twin_cat()

        def with_timeout(self: "TwinCAT._Builder", timeout: timedelta) -> "TwinCAT._Builder":
            """Set timeout.

            Arguments:
            ---------
                timeout: Timeout
            """
            self._builder = LinkTwinCAT().link_twin_cat_with_timeout(self._builder, int(timeout.total_seconds() * 1000 * 1000 * 1000))
            return self

        def _link_builder_ptr(self: "TwinCAT._Builder") -> LinkBuilderPtr:
            return LinkTwinCAT().link_twin_cat_into_builder(self._builder)  # pragma: no cover

        def _resolve_link(self: "TwinCAT._Builder", _ptr: ControllerPtr) -> "TwinCAT":
            return TwinCAT(Base().link_get(_ptr))

    def __init__(self: "TwinCAT", ptr: LinkPtr) -> None:
        super().__init__(ptr)  # pragma: no cover

    @staticmethod
    def builder() -> _Builder:
        """Create TwinCAT link builder."""
        return TwinCAT._Builder()


class RemoteTwinCAT(Link):
    """Link for remote TwinCAT3 server via [ADS](https://github.com/Beckhoff/ADS) library."""

    class _Builder(LinkBuilder):
        _builder: LinkRemoteTwinCATBuilderPtr

        def __init__(self: "RemoteTwinCAT._Builder", server_ams_net_id: str) -> None:
            self._builder = _validate_ptr(LinkTwinCAT().link_remote_twin_cat(server_ams_net_id.encode("utf-8")))

        def with_server_ip(self: "RemoteTwinCAT._Builder", ip: str) -> "RemoteTwinCAT._Builder":
            """Set server IP address.

            Arguments:
            ---------
                ip: Server IP address
            """
            self._builder = LinkTwinCAT().link_remote_twin_cat_with_server_ip(self._builder, ip.encode("utf-8"))
            return self

        def with_client_ams_net_id(self: "RemoteTwinCAT._Builder", ams_net_id: str) -> "RemoteTwinCAT._Builder":
            """Set client AMS Net ID.

            Arguments:
            ---------
                ams_net_id: Client AMS Net ID
            """
            self._builder = LinkTwinCAT().link_remote_twin_cat_with_client_ams_net_id(self._builder, ams_net_id.encode("utf-8"))
            return self

        def with_timeout(self: "RemoteTwinCAT._Builder", timeout: timedelta) -> "RemoteTwinCAT._Builder":
            """Set timeout.

            Arguments:
            ---------
                timeout: Timeout
            """
            self._builder = LinkTwinCAT().link_remote_twin_cat_with_timeout(self._builder, int(timeout.total_seconds() * 1000 * 1000 * 1000))
            return self

        def _link_builder_ptr(self: "RemoteTwinCAT._Builder") -> LinkBuilderPtr:
            return LinkTwinCAT().link_remote_twin_cat_into_builder(self._builder)  # pragma: no cover

        def _resolve_link(self: "RemoteTwinCAT._Builder", _ptr: ControllerPtr) -> "RemoteTwinCAT":
            return RemoteTwinCAT(Base().link_get(_ptr))

    def __init__(self: "RemoteTwinCAT", ptr: LinkPtr) -> None:
        super().__init__(ptr)  # pragma: no cover

    @staticmethod
    def builder(server_ams_net_id: str) -> _Builder:
        """Create TwinCAT link builder.

        Arguments:
        ---------
            server_ams_net_id: Server AMS Net ID
        """
        return RemoteTwinCAT._Builder(server_ams_net_id)

File: dotnet/cs/src/Link/TwinCAT.cs
/*
 * File: TwinCAT.cs
 * Project: Link
 * Created Date: 20/08/2023
 * Author: Shun Suzuki
 * -----
 * Last Modified: 06/12/2023
 * Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
 * -----
 * Copyright (c) 2023 Shun Suzuki. All rights reserved.
 *
 */

using AUTD3Sharp.Internal;
using AUTD3Sharp.NativeMethods;
using System;
using System.Net;

namespace AUTD3Sharp.Link
{
    /// <summary>
    /// Link using TwinCAT3
    /// </summary>
    public sealed class TwinCAT
    {
        public sealed class TwinCATBuilder : ILinkBuilder<TwinCAT>
        {
            private LinkTwinCATBuilderPtr _ptr;

            internal TwinCATBuilder()
            {
                _ptr = NativeMethodsLinkTwinCAT.AUTDLinkTwinCAT();
            }

            public TwinCATBuilder WithTimeout(TimeSpan timeout)
            {
                _ptr = NativeMethodsLinkTwinCAT.AUTDLinkTwinCATWithTimeout(_ptr, (ulong)(timeout.TotalMilliseconds * 1000 * 1000));
                return this;
            }

            LinkBuilderPtr ILinkBuilder<TwinCAT>.Ptr()
            {
                return NativeMethodsLinkTwinCAT.AUTDLinkTwinCATIntoBuilder(_ptr);
            }

            TwinCAT ILinkBuilder<TwinCAT>.ResolveLink(LinkPtr ptr)
            {
                return new TwinCAT();
            }
        }

        public static TwinCATBuilder Builder()
        {
            return new TwinCATBuilder();
        }
    }

    /// <summary>
    /// Link for remote TwinCAT3 server via <see href="https://github.com/Beckhoff/ADS">ADS</see> library
    /// </summary>
    public sealed class RemoteTwinCAT
    {
        public sealed class RemoteTwinCATBuilder : ILinkBuilder<RemoteTwinCAT>
        {
            private LinkRemoteTwinCATBuilderPtr _ptr;

            /// <summary>
            /// Constructor
            /// </summary>
            /// <param name="serverAmsNetId">Server AMS Net ID</param>
            /// <exception cref="AUTDException"></exception>
            public RemoteTwinCATBuilder(string serverAmsNetId)
            {
                var serverAmsNetIdBytes = System.Text.Encoding.UTF8.GetBytes(serverAmsNetId);
                unsafe
                {
                    fixed (byte* ap = &serverAmsNetIdBytes[0])
                    {
                        _ptr = NativeMethodsLinkTwinCAT.AUTDLinkRemoteTwinCAT(ap).Validate();
                    }
                }
            }

            /// <summary>
            /// Set server IP address
            /// </summary>
            /// <param name="serverIp"></param>
            /// <returns></returns>
            public RemoteTwinCATBuilder WithServerIp(IPAddress serverIp)
            {
                var serverIpBytes = serverIp.GetAddressBytes();
                unsafe
                {
                    fixed (byte* ap = &serverIpBytes[0])
                        _ptr = NativeMethodsLinkTwinCAT.AUTDLinkRemoteTwinCATWithServerIP(_ptr, ap);
                }

                return this;
            }

            /// <summary>
            /// Set client AMS Net ID
            /// </summary>
            /// <param name="clientAmsNetId"></param>
            /// <returns></returns>
            public RemoteTwinCATBuilder WithClientAmsNetId(string clientAmsNetId)
            {
                var clientAmsNetIdBytes = System.Text.Encoding.UTF8.GetBytes(clientAmsNetId);
                unsafe
                {
                    fixed (byte* ap = &clientAmsNetIdBytes[0])
                        _ptr = NativeMethodsLinkTwinCAT.AUTDLinkRemoteTwinCATWithClientAmsNetId(_ptr, ap);
                }
                return this;
            }

            public RemoteTwinCATBuilder WithTimeout(TimeSpan timeout)
            {
                _ptr = NativeMethodsLinkTwinCAT.AUTDLinkRemoteTwinCATWithTimeout(_ptr, (ulong)(timeout.TotalMilliseconds * 1000 * 1000));
                return this;
            }

            LinkBuilderPtr ILinkBuilder<RemoteTwinCAT>.Ptr()
            {
                return NativeMethodsLinkTwinCAT.AUTDLinkRemoteTwinCATIntoBuilder(_ptr);
            }

            RemoteTwinCAT ILinkBuilder<RemoteTwinCAT>.ResolveLink(LinkPtr ptr)
            {
                return new RemoteTwinCAT();
            }
        }

        public static RemoteTwinCATBuilder Builder(string serverAmsNetId)
        {
            return new RemoteTwinCATBuilder(serverAmsNetId);
        }
    }
}
